### DISCUSSION TITLE
升级WebPack5后希望继续使用crypto和buffer的问题

### DISCUSSION THREAD

**Discussion by zspggp** [NervJS/taro#13657] [Q & A] [answered]
### 这个特性解决了什么问题？
如题，WebPack5无法直接使用node的原生模块crypto和buffer等，请问有没有测试可行的解决方案呢？

### 这个 API 长什么样？
期望可以通过配置，使用WebPack5打包也可以继续使用crypto和buffer等。

**Comment by yoyo837** [comment, marked as answer]
```js
new webpack.ProvidePlugin({
  // 浏览器兼容 webpack 4 的 Buffer
  Buffer: ['buffer', 'Buffer'],
  // 浏览器兼容 webpack 4 的 process
  'process.browser': JSON.stringify(true),
}),
```

**Reply by pisyell** [reply]
这个怎么配置，具体配在哪里？

**Reply by Solyna** [reply]
请问下在Taro里怎么配置？

**Reply by yoyo837** [reply]
webpack配置里, 请尽力先阅读一下文档, 实在没精力快速响应每个诉求. 连续加班一周了, 比996还惨

**Reply by Solyna** [reply]
不好意思打扰了，我是这样配置的：但是经过验证，在支付宝小程序没报错，在微信小程序还是报错 Can't find variable: Buffer
   webpackChain (chain) {
      chain.plugin('ProvidePlugin').use(ProvidePlugin, [{
        process: ['process/browser.js'] ,
        Buffer: ['buffer', 'Buffer'],
      }]);
    }

**Comment by yoyo837** [comment]
`crypto` 可以 `fallback` 到 `crypto-browserify`

**Reply by pisyell** [reply]
> `crypto` 可以 `fallback` 到 `crypto-browserify`

webpack5模式又如何fallback？

**Reply by yoyo837** [reply]
请直接查阅一下webpack5文档.

**Reply by pisyell** [reply]
> 请直接查阅一下webpack5文档.

webpack5我会啊 关键是在taro中怎么配置？在哪配置webpack5的东西 以及fallback能生效不

**Reply by yoyo837** [reply]
试试 webpackChain

**Reply by Solyna** [reply]
配置了但是没有用啊。。急救

 webpackChain(chain, webpack) {
      chain
        .plugin("ProvidePlugin")
        .use(require.resolve("webpack/lib/ProvidePlugin"), [
          {
            Buffer: ["buffer", "Buffer"],
            // process: ['process/browser.js'],
          },
        ]);
      chain.merge({
        resolve: {
          fallback: {
            buffer: require.resolve("buffer"),
            // process: require.resolve('process/browser'),
          },
        },
      });
    },

**Comment by panhezeng** [comment]
```js
      chainWebpack(chain, {isServer, isClient}) {
        chain.resolve.alias.set('@', path.resolve(__dirname, 'src'));
        .set("crypto", "crypto-browserify")
        .set("stream", "stream-browserify")
        .set("process", "process/browser.js")
        .set("buffer", "buffer");

        chain
           .plugin("provide")
           .use(require.resolve("webpack/lib/ProvidePlugin"), [
             {
               crypto: "crypto-browserify",
               stream: "stream-browserify",
               process: "process/browser.js",
               Buffer: ["buffer", "Buffer"],
            },
           ]);
```

**Comment by UphkcRNqmafQWcSSSSSS** [comment]
```ts
// npm i -D node-polyfill-webpack-plugin
module.exports = {
 plugins: ['node-polyfill-webpack-plugin']
}
```

**Reply by Solyna** [reply]
会报错：
![image](https://github.com/NervJS/taro/assets/40814530/c0b37789-22ae-45a3-ace2-3f646821dfbc)

**Reply by BLJJ-DBLD** [reply]
```
// config/index.ts 中配置
import NodePolyfillPlugin from "node-polyfill-webpack-plugin";

mini: {
    webpackChain(chain) {
      chain.plugin("node-polyfill").use(
        new NodePolyfillPlugin({
          includeAliases: ["Buffer"],
        })
      );
    },
  },
```

我使用是生效了，没报错了

**Reply by brezhal** [reply]
mini: {
    webpackChain(chain) {
      chain.plugin("node-polyfill").use(
        new NodePolyfillPlugin({
          additionalAliases: ["Buffer", 'process'],
        })
      );
    },
}
